### DISCUSSION TITLE
What is the best approach to use map offline?

### DISCUSSION THREAD

**Discussion by tashi146** [fleaflet/flutter_map#1250] [General]
I have spend couple of day trying to download map using flutter_map and result was upset. Here is what I tried 

1.  **tilemill** (Which is depreciated and not working)
2. **MapBox** which is confusing and there is no up to date tutorial
3. **MBTileProvider** (also deprecated and removed from the package) 

Please do let me know best approach or up to date resource where I can achieve offline map.

**Comment by JaffaKetchup** [comment]
Hi there,
What kind of offline do you need? A more dynamic approach where tiles are saved on the go or in bulk downloading, or a way you can ship a set of tiles with each app?
Thanks!

**Reply by tashi146** [reply]
Woh, thanks for the quick respond. Yes It is bulk downloading. I am looking for something similar to this https://github.com/fleaflet/flutter_map/issues/223. But I am struck with generating .mbtiles and extracting to png.     

**[Here is what I am expecting]**    
User should able to download map by pressing download button and use map offline.   

**[My Question]**
Could you let me know 
1. How can we generate .mbtiles since tilemill is depreciated and I don't see any option to create .mbtiles in MapBox as well
2. I got few demo .mbtiles somehow but not able to convert to .png using https://github.com/alfanhui/mbtilesToPngs

 If this is not the best, What are the alternative that you can suggest for my requirement.   

Regards

**Reply by JaffaKetchup** [reply]
No problem, downloading is my 'speciality'!

In this case you are probably not looking to use .mbtiles, as these cannot be generated on the user's device easily. In any case, creating a set of .mbtiles is notoriously high maintenance and difficult if you don't have a super-computer. You probably got stuck on your second step because you were using vector .mbtiles (more common and easy to create) which need rendering - which is the tricky bit.

My recommendation is, to provide bulk downloading, you'll either need to create your own method, or use my plugin for flutter_map '[flutter_map_tile_caching](https://github.com/JaffaKetchup/flutter_map_tile_caching)'. Note that many providers (such as OpenStreetMap) will forbid bulk downloading from their server, as it consumes lots of resources. My plugin, if you choose to use it, has an example app (ready for Android, will need some more effort for iOS) that can show you how bulk downloading might look from a UI perspective and a developer's perspective.

If you need more help with my plugin, please drop me an issue over there (it's not the easiest to setup, but easy once you understand it). If you want more help here, because I've probably misunderstood something, I'm happy to help more!

**Reply by tashi146** [reply]
Great, But I need desktop support as well (at least window support), anyway thanks I will check it out.    

**Just to give you a broad idea.**
Actually, I wanted to implement[ ArcGIS offline](https://doc.arcgis.com/en/arcgis-online/manage-data/take-maps-offline.htm). But coming to your package I end up with .mbtiles and so on.   

**[Here come the annoying questions]**   

1. So, is it possible to implement ArcGIS offline with your package?   
2. If so can you provide me with some resource

I have tried [flutter_map_arcgis](https://github.com/khankhulgun/flutter_map_arcgis) and raised a similar issue [here](https://github.com/khankhulgun/flutter_map_arcgis/issues/18) as well, Let me know what you think.   

Thanks for the awesome support

**Comment by JaffaKetchup** [comment]
Not really sure what ArcGIS is, but on that page:

 - .mbtiles is:
    > Create map areas—Map authors can create map areas to package data for specific areas ahead of time. Map areas speed up and simplify the map download process for field workers.

 - my package/suggestion is:
    > On demand—Field users can download any portion of the map whenever they need it. The field user specifies the area they want to take offline along with the details of the basemap required.

Having a quick skim read about creating an offline area, it seems it might output the .mbtiles you are talking about. This, however, is probably a vector format .mbtiles. How quick does it export (if this is what you were doing), as that will reveal it's format? But the actual process itself (of using shapes to select a region of a map to download) is very similar to what my package can provide.

**Reply by tashi146** [reply]
Yes, I am expecting a huge size maybe around 1.5GB of space that needs to download from server to device and then use the downloaded map on mobile devices (Still not sure whether this is a good approach or not). 

It might be costly but downloading offline is a one time. Once downloaded, users can use maps without any network connection. Having said that I will find out more about ArcGIS and play around with flutter_map_tile_caching. and see if I can get anything.

Thanks again

**Reply by JaffaKetchup** [reply]
No problem :)

**Reply by tashi146** [reply]
I am familiar with the platform issue that you are getting and I have noticed it in some of your other packages as well. You need to explicitly defined platform support in pubspec.yaml.   
```
platforms:
  android:
  ios:
  web:
  windows:
```   
Like so. Than you are good to go

**Reply by JaffaKetchup** [reply]
Ah, I see. Didn't know about that, thanks!
